server/zio-http/src/main/scala/sttp/tapir/server/zhttp/ZHttpInterpreter.scala

```diff
@@ -0,0 +1,140 @@
+      }
+    case RawBodyType.ByteArrayBody =>
+      request.data.content match {
+        case HttpContent.Complete(data) => Task.succeed(data.getBytes())
+        case HttpContent.Chunked(data) => data.fold("")(_ + _).map(_.getBytes)
+      }
+    case RawBodyType.ByteBufferBody => ???
+    case RawBodyType.InputStreamBody => ???
+    case RawBodyType.FileBody => ???
+    case RawBodyType.MultipartBody(partTypes, defaultType) => ???
+  }
+
+  def toBytes(charset: Charset): ZTransducer[Any, Nothing, String, Byte] =
+    ZTransducer.fromPush {
+      case Some(chunk) => ZIO.succeed(chunk.flatMap(_.getBytes(charset)))
+      case None => ZIO.succeed(Chunk.empty)
+    }
+
+  val stream: Stream[Throwable, Byte] = request.data.content match {
+    case HttpContent.Complete(data) => ZStream.fromIterator(data.getBytes.iterator)
+    case HttpContent.Chunked(data) => data.transduce(toBytes(StandardCharsets.UTF_8))
+  }
+
+  def toStream(): streams.BinaryStream = stream.asInstanceOf[streams.BinaryStream] //TODO: this is weird we need a aggressive type cast
+}
+
+class ZHttpToResponseBody extends ToResponseBody[Stream[Nothing, String], ZioStreams] {
+  val streams: capabilities.Streams[ZioStreams] = ZioStreams
+
+  override def fromRawValue[R](v: R, headers: HasHeaders, format: CodecFormat, bodyType: RawBodyType[R]): Stream[Nothing, String] =
+    rawValueToEntity(bodyType, v)
+
+  override def fromStreamValue(v: streams.BinaryStream, headers: HasHeaders, format: CodecFormat, charset: Option[Charset]): Stream[Nothing, String] =
+    v.asInstanceOf[Stream[Throwable, Byte]].transduce(Transducer.utf8Decode).refineOrDie(PartialFunction.empty) //TODO
+
+  override def fromWebSocketPipe[REQ, RESP](pipe: streams.Pipe[REQ, RESP], o: WebSocketBodyOutput[streams.Pipe[REQ, RESP], REQ, RESP, _, ZioStreams]): Stream[Nothing, String] =
+    Stream.empty //TODO
+
+  private def rawValueToEntity[CF <: CodecFormat, R](bodyType: RawBodyType[R], r: R): Stream[Nothing, String] = {
+    bodyType match {
+      case RawBodyType.StringBody(charset) =>
+        ZStream.succeed(r.toString)
+      case _ =>
+        ZStream.empty
+    }
+  }
+}
+
+object ZHttpInterpreter {
+
+  implicit def monadError[R]: MonadError[RIO[R, *]] = new MonadError[RIO[R, *]] {
+    def unit[T](t: T): RIO[R, T] = ZIO.succeed(t)
+    def map[T, T2](fa: RIO[R, T])(f: T => T2): RIO[R, T2] = fa.map(f)
+    def flatMap[T, T2](fa: RIO[R, T])(f: T => RIO[R, T2]): RIO[R, T2] = fa.flatMap(f)
+    def error[T](t: Throwable): RIO[R, T] = ZIO.fail(t)
+    protected def handleWrappedError[T](rt: RIO[R, T])(h: PartialFunction[Throwable, RIO[R, T]]): RIO[R, T] = rt.catchSome(h)
+    def ensure[T](f: RIO[R, T], e: => RIO[R, Unit]): RIO[R, T] = f.ensuring(e.orDie)
+  }
+
+  private def sttpToZHttpHeader(header: SttpHeader): ZHttpHeader =
+    ZHttpHeader(header.name, header.value)
+
+  def convert[I, O, R](route: Endpoint[I, Throwable, O, ZioStreams])(logic: I => RIO[R, O]): Http[R, Throwable] = {
```

Comment:
maybe we can call `toHttp`? that would follow the convention in other interpreters

Reply:
Yes, also planning to support multiple endpoints in one go.